synapse/api/errors.py

```diff
@@ -30,7 +31,17 @@
 logger = logging.getLogger(__name__)
 
 
-class Codes:
+class Codes(str, Enum):
+    """
+    All known error codes, as an enum of strings.
+    """
+
+    def __str__(self) -> str:
+        return self.value
+
+    def __repr__(self) -> str:
+        return repr(self.value)
```

Comment:
What are these for? I can't see a difference in behaviour between this version and how it works on `develop`:

This branch:

```python
>>> from synapse.api.errors import Codes
>>> Codes.MISSING_TOKEN
'M_MISSING_TOKEN'
>>> repr(Codes.MISSING_TOKEN)
"'M_MISSING_TOKEN'"
>>> 
```

`develop`:

```python
>>> from synapse.api.errors import Codes
>>> str(Codes.MISSING_TOKEN)
'M_MISSING_TOKEN'
>>> repr(Codes.MISSING_TOKEN)
"'M_MISSING_TOKEN'"
>>> 
```

Reply:
Yes, that's to allow backwards compatibility.

Without these, `Codes.foo` becomes a `Codes`, which won't covert to a `str`, so we break all the sites that call e.g. `Codes.MISSING_TOKEN`.

Reply:
> Without these, Codes.foo becomes a Codes, which won't covert to a str

But with current develop I'm able to do `str(Codes.MISSING_TOKEN)` and it returns the same thing, so I'm not sure what you're talking about. Can you show me an example of code that wouldn't work without this change?

Reply:
My bad, it doesn't seem necessary.

Reply:
Are you sure?
```py
>>> from enum import Enum
>>> class Codes(str, Enum):
...     MISSING_TOKEN = "M_MISSING_TOKEN"
...
>>> f"{Codes.MISSING_TOKEN}"
'M_MISSING_TOKEN'
>>> "%s" % (Codes.MISSING_TOKEN,)
'Codes.MISSING_TOKEN'
>>> str(Codes.MISSING_TOKEN)
'Codes.MISSING_TOKEN'

```